### ISSUE TITLE
Encryption key must be CRYPTO_STREAM_KEYBYTES bytes long

### ISSUE THREAD

**Issue by ajaybairagi** [paragonie/halite#98]
Hello,

I am trying to solve Error:  `Encryption key must be CRYPTO_STREAM_KEYBYTES bytes long`  
$secret = new HiddenString(env('nako4id94oln4p0s'));
 $salt = random_bytes(16);
$encryptionKey = KeyFactory::deriveEncryptionKey($secret, $salt);

it is working fine but not work on server.
also I tried   $salt = 41ca73cffcbf4942
but not work. I am using in laravel lumen

**Comment by obozaxy** [comment]
It needs to be 32 bytes not 16.

**Comment by paragonie-scott** [comment]
What version of Halite? Using the `deriveEncryptionKey()` method should work with any length password.

**Comment by ajaybairagi** [comment]
Latest version.
It is working on local system fine but not run on server

**Comment by ajaybairagi** [comment]
I am using Halite 4.1 version 

I want to call password and salt from .env file.
Case 1 ) 
PASSWD =nako4id94oln4p0s
SALT=41ca73cffcbf494241ca73cffcbf4942
give below error
"error": "Expected 16 bytes, got 32"

Case 2) 

PASSWD =nako4id94oln4p0s
SALT=41ca73cffcbf4942
give below error
Error: "Encryption key must be CRYPTO_STREAM_KEYBYTES bytes long"

**Comment by paragonie-scott** [comment]
`Hex::decode()` your salt. It must be 16 bytes. That's a libsodium error, not a Halite error.

**Comment by paragonie-scott** [comment]
```
$ php -a
Interactive mode enabled

php > require "vendor/autoload.php";
php > $passwd = new ParagonIE\Halite\HiddenString('nako4id94oln4p0s');
php > $salt = ParagonIE\ConstantTime\Hex::decode('41ca73cffcbf494241ca73cffcbf4942');
php > $key = ParagonIE\Halite\KeyFactory::deriveEncryptionKey($passwd, $salt);
php > $save = ParagonIE\Halite\KeyFactory::export($key);
php > var_dump($save);
object(ParagonIE\Halite\HiddenString)#5 (2) {
  ["internalStringValue"]=>
  string(1) "*"
  ["attention"]=>
  string(82) "If you need the value of a HiddenString, invoke getString() instead of dumping it."
}
php > var_dump($save->getString());
string(200) "3140040047a89f22e51976bd6f4d952985241e3b98986ebcee91a82b921b9e35d18cfa2a1bc61cadfb57718b70da3d3fc11da0262971a81ed2db7b4ca841e5e49e69c063a627ff33c44277bd24468740ffbc9503066cec6cf77f05d50e0e45576fdd0f57"
```

**Comment by rkeet** [comment]
[The message reported is at this line](https://github.com/paragonie/halite/blob/7fef56c1f540d8fc0023cb0f3d236626db8d79b3/src/Symmetric/EncryptionKey.php#L29)

```
        if (CryptoUtil::safeStrlen($keyMaterial->getString()) !== \Sodium\CRYPTO_STREAM_KEYBYTES) {
            throw new InvalidKey(
                'Encryption key must be CRYPTO_STREAM_KEYBYTES bytes long'
            );
        }
```

Could be rewritten to:

```
        if (CryptoUtil::safeStrlen($keyMaterial->getString()) !== \Sodium\CRYPTO_STREAM_KEYBYTES) {
            throw new InvalidKey(
                'Encryption key must be ' . \Sodium\CRYPTO_STREAM_KEYBYTES . ' bytes long'
            );
        }
```

This happens more throughout the code, such as [here](https://github.com/paragonie/halite/blob/7fef56c1f540d8fc0023cb0f3d236626db8d79b3/src/Asymmetric/EncryptionPublicKey.php#L31),  [here](https://github.com/paragonie/halite/blob/7fef56c1f540d8fc0023cb0f3d236626db8d79b3/src/Asymmetric/EncryptionSecretKey.php#L29), [here](https://github.com/paragonie/halite/blob/7fef56c1f540d8fc0023cb0f3d236626db8d79b3/src/Asymmetric/SignaturePublicKey.php#L31) and frankly quite a few more locations. 

Just my 2 cents on this issue. Could also do a quick search for the constant and find the value you need.

**Comment by paragonie-scott** [comment]
I'll put the value inside parentheses after the constant name. (This won't be backported into unsupported versions of Halite.)

**Comment by paragonie-security** [comment]
This will be fixed in v5.0.0
